numgrid/AtomGrid.cpp

```diff
@@ -0,0 +1,284 @@
+{
+    int min_num_angular_points_closest =
+        get_closest_num_angular(min_num_angular_points);
+    int max_num_angular_points_closest =
+        get_closest_num_angular(max_num_angular_points);
+
+    double *angular_x = new double[MAX_ANGULAR_ORDER * MAX_ANGULAR_GRID];
+    double *angular_y = new double[MAX_ANGULAR_ORDER * MAX_ANGULAR_GRID];
+    double *angular_z = new double[MAX_ANGULAR_ORDER * MAX_ANGULAR_GRID];
+    double *angular_w = new double[MAX_ANGULAR_ORDER * MAX_ANGULAR_GRID];
+
+    for (int i = get_angular_order(min_num_angular_points_closest);
+         i <= get_angular_order(max_num_angular_points_closest);
+         i++)
+    {
+        int angular_off = i * MAX_ANGULAR_GRID;
+        ld_by_order(lebedev_table[i],
+                    &angular_x[angular_off],
+                    &angular_y[angular_off],
+                    &angular_z[angular_off],
+                    &angular_w[angular_off]);
+    }
+
+    // obtain radial parameters
+    double r_inner = get_r_inner(radial_precision,
+                                 alpha_max * 2.0); // factor 2.0 to match DIRAC
+    double h = std::numeric_limits<float>::max();
+    double r_outer = 0.0;
+    for (int l = 0; l <= max_l_quantum_number; l++)
+    {
+        if (alpha_min[l] > 0.0)
+        {
+            r_outer =
+                std::max(r_outer,
+                         get_r_outer(radial_precision,
+                                     alpha_min[l],
+                                     l,
+                                     4.0 * get_bragg_angstrom(proton_charge)));
+            NUMGRID_ASSERT(r_outer > r_inner);
+            h = std::min(h,
+                         get_h(radial_precision, l, 0.1 * (r_outer - r_inner)));
+        }
+    }
+    NUMGRID_ASSERT(r_outer > h);
+
+    num_grid_points = 0;
+    num_radial_grid_points = 0;
+
+    double rb = get_bragg_angstrom(proton_charge) /
+                (5.0 * 0.529177249); // factors match DIRAC code
+    double c = r_inner / (exp(h) - 1.0);
+    int num_radial = int(log(1.0 + (r_outer / c)) / h);
+    for (int irad = 0; irad < num_radial; irad++)
+    {
+        double radial_r = c * (exp((irad + 1) * h) - 1.0);
+        double radial_w = (radial_r + c) * radial_r * radial_r * h;
+
+        radial_atom_grid_r_au.push_back(radial_r);
+        radial_atom_grid_w.push_back(radial_w);
+        num_radial_grid_points++;
+
+        int num_angular = max_num_angular_points_closest;
+        if (radial_r < rb)
+        {
+            num_angular = static_cast<int>(max_num_angular_points_closest *
+                                           (radial_r / rb));
+            num_angular = get_closest_num_angular(num_angular);
+            if (num_angular < min_num_angular_points_closest)
+                num_angular = min_num_angular_points_closest;
+        }
+
+        int angular_off = get_angular_order(num_angular) * MAX_ANGULAR_GRID;
+
+        for (int iang = 0; iang < num_angular; iang++)
```

Comment:
You should use `std::copy` instead of `push_back` and a loop.

Reply:
Thanks!